ambari-server/src/main/python/ambari_server/setupSso.py

```diff
@@ -113,15 +125,101 @@ def populateJwtAudiences(options, properties):
 
   properties.process_pair(JWT_AUDIENCES, audiences)
+  
+def get_eligible_services(properties, admin_login, admin_password, cluster_name):
+  url = get_ambari_server_api_base(properties) + URL_TO_FETCH_SERVICES_ELIGIBLE_FOR_SSO.replace(":CLUSTER_NAME", cluster_name)
+  admin_auth = base64.encodestring('%s:%s' % (admin_login, admin_password)).replace('\n', '')
+  request = urllib2.Request(url)
+  request.add_header('Authorization', 'Basic %s' % admin_auth)
+  request.add_header('X-Requested-By', 'ambari')
+  request.get_method = lambda: 'GET'
+
+  services = []
+  sys.stdout.write('\nFetching SSO enabled services')
+  numOfTries = 0
+  request_in_progress = True
+  while request_in_progress:
+    numOfTries += 1
+    if (numOfTries == 60):
+      raise FatalException(1, "Could not fetch eligible services within a minute; giving up!")
+    sys.stdout.write('.')
+    sys.stdout.flush()
+
+    try:
+      with closing(urllib2.urlopen(request)) as response:
+        response_status_code = response.getcode()
+        if response_status_code != 200:
+          request_in_progress = False
+          err = 'Error while fetching eligible services. Http status code - ' + str(response_status_code)
+          raise FatalException(1, err)
+        else:
+            response_body = json.loads(response.read())
+            items = response_body['items']
+            if len(items) > 0:
+              for item in items:
+                services.append(item['ServiceInfo']['service_name'])
+            if not items:
+              time.sleep(1)
+            else:
+              request_in_progress = False
+
+    except Exception as e:
+      request_in_progress = False
+      err = 'Error while fetching eligible services. Error details: %s' % e
+      raise FatalException(1, err)
+  if (len(services) == 0):
+    sys.stdout.write('\nThere is no SSO enabled services found\n')
+  else:
+    sys.stdout.write('\nFound SSO enabled services: {0}\n'.format(', '.join(str(s) for s in services)))
+  return services
+
+def get_services_requires_sso(options, properties, admin_login, admin_password):
+  if not options.sso_enabled_services:
+    configure_for_all_services = get_YN_input("Use SSO for all services [y/n] (n): ", False)
+    if configure_for_all_services:
+      services = WILDCARD_FOR_ALL_SERVICES
+    else:
+      cluster_name = get_cluster_name(properties, admin_login, admin_password)
+      eligible_services = get_eligible_services(properties, admin_login, admin_password, cluster_name)
+      services = ''
+      for service in eligible_services:
+        question = "Use SSO for {0} [y/n] (y): ".format(service)
+        if get_YN_input(question, True):
+          if len(services) > 0:
+            services = services + ", "
+          services = services + service
+  else:
+    services = options.sso_enabled_services
+
+  return services
+
+
+def update_sso_conf(properties, services, admin_login, admin_password):
+  sso_configuration_properties = {}
```

Comment:
Add `ambari.sso.manage_services` to the properties indicating "true" to trigger Ambari to manage services, or "false" to have Ambari do nothing.

Reply:
ok

Reply:
Done